### ISSUE TITLE
First stage ops

### ISSUE THREAD

**Issue by raver119** [deeplearning4j/nd4j#2146]
List of ops required for ND4j/libND4j for first stage:

- [x] Local Response Normalization
- [x] Conv2D
- [x] MaxPooling2D
- [x] AvgPooling2D
- [x] Softmax (just a wrapper over existing op)
- [x] BiasAdd (just a wrapper over addiRowVector)
- [x] MatMul (probably exists)

**Comment by raver119** [comment]
List of ops for next stage:

- [ ] deconv 2d
- [ ] separable conv2d
- [ ] batched gemm wrapper
- [ ] BatchNorm
- [ ] Bonus: AdaptiveAvgPooling, AdaptiveMaxPooling, MaxUnpooling2D

List of logic improvements:

- [ ] Graph in Graph. Nodes can be graphs too.
- [ ] Aggregations overhaul. To be invented yet, but aggregations might use some love.
- [ ] Stash for backprop pass in Workspace
